File: graph/topological-sort.py
import sys
from collections import defaultdict


try:
    sys.stdin = open(sys.path[0] + '\\input.txt', 'r')
    sys.stdout = open(sys.path[0] + '\\output.txt', 'w')
except FileNotFoundError:
    pass


def dfs(u):
    visited[u] = True

    for i in graph[u]:
        if not visited[i]:
            dfs(i)
    
    stack.append(u)


v, e = list(map(int, input().split()))
graph = defaultdict(list)

for i in range(e):
    a, b = list(map(int, input().split()))
    graph[a].append(b)

visited = [False] * v
stack = []

for i in range(v):
    if not visited[i]:
        dfs(i)

print(stack[::-1])

d = {i: chr(i + 65) for i in range(v)}
for i in reversed(stack):
    print(d[i], end=' ')

File: graph/topological-sort.cpp
#include <bits/stdc++.h>

#define MAX 1000005
#define ll long long
#define watch(x) cout << (#x) << " is " << (x) << endl

using namespace std;

vector<int> s;

void dfs(vector<vector<int>> &adj, bool visited[], int u) {
	visited[u] = true;

	for (auto i = adj[u].begin(); i != adj[u].end(); i++) {
		if (!visited[*i]) {
			dfs(adj, visited, *i);
		}
	}

	s.push_back(u);
}

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);

    #ifndef ONLINE_JUDGE
    freopen("input.txt", "r", stdin);
    freopen("output.txt", "w", stdout);
    #endif

	int v, e;
	cin >> v >> e;

	vector<vector<int>> adj;
	adj.assign(v, vector<int>());

	int a, b;
	for (int i = 0; i < e; i++) {
		cin >> a >> b;
		adj[a].push_back(b);
	}

	bool visited[v];
	fill(visited, visited + v, false);

	for (int i = 0; i < v; i++) {
		if (!visited[i]) {
			dfs(adj, visited, i);
		}
	}

	for (auto i = s.rbegin(); i != s.rend(); i++) {
		cout << *i << " ";
	}

    return 0;
}